server/auth/store.go

```diff
@@ -388,6 +388,8 @@ func (as *authStore) Recover(be AuthBackend) {
 		as.tokenProvider.enable()
 	}
 	as.enabledMu.Unlock()
+
+	as.rangePermCache = make(map[string]*unifiedRangePermissions)
```

Comment:
Two comments:
1. Suggest to call [clearCachedPerm](https://github.com/etcd-io/etcd/blob/3a239f7b3b02964deb4f2dd4489856e01fbb9906/server/auth/range_perm_cache.go#L128);
2. There is a potential race condition, some requests (such as [v3_server.go#L128](https://github.com/etcd-io/etcd/blob/3a239f7b3b02964deb4f2dd4489856e01fbb9906/server/etcdserver/v3_server.go#L128) and [watch.go#L235](https://github.com/etcd-io/etcd/blob/3a239f7b3b02964deb4f2dd4489856e01fbb9906/server/etcdserver/api/v3rpc/watch.go#L235)  ) coming from API (outside of the applying workflow) may be accessing the `rangePermCache` concurrently.  It seems that we need to add a lock to protect it.

Reply:
+1. Good catch. 

In general the existing locking strategy seems week here:
 - The access to this cache is protected by `readtx.Lock()`
 -  `readTx` is owned by backend that we are swapping here.
 -  so seems theoretically possible that there would be 2 transactions concurrently accessing/modifying the cache... one having tx on old backend, the other the new backend.   Closing of the old backend is fully asynchronic:
https://github.com/etcd-io/etcd/blob/2e034d2b65b2c6c7b2c1186087c84d4e98596479/server/etcdserver/server.go#L1000-L1008
- Thus seems cache should have its own lock instead of piggybacking on transaction lock.

Reply:
> Thus seems cache should have its own lock instead of piggybacking on transaction lock

Exactly! Please note that the cache can't be protected by the readtx.Lock(), because a batchTx and a readRx can execute concurrently.

Reply:
Thanks for pointing out that! I think it's an independent issue, let me open a dedicated PR for it.

Reply:
Opened an independent PR here: https://github.com/etcd-io/etcd/pull/13954 It's great if you can review.